br/pkg/lightning/backend/external/kv_buf.go

```diff
@@ -0,0 +1,72 @@
+// Copyright 2023 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package external
+
+import "github.com/docker/go-units"
+
+// DefaultBlockSize is the default block size for preAllocKVBuf.
+const DefaultBlockSize = 16 * units.MiB
+
+// preAllocKVBuf pre allocates a large buffer of limit memLimit to reduce memory
+// allocation, all space in this buffer will be reused when reset.
+type preAllocKVBuf struct {
+	blocks      [][]byte
+	blockSize   int
+	curBlock    []byte
+	curBlockIdx int
+	curIdx      int
+}
+
+func newPreAllocKVBuf(memLimit uint64, blockSize int) *preAllocKVBuf {
```

Comment:
it looks similar to membuf.Buffer 🤔 not sure we can reuse it

Reply:
fields are similar, but not the code, will leave it for now